proposals/lip-0053.md

```diff
@@ -34,127 +34,262 @@ Identifies the chain sending the cross-chain update. Only cross-chain update tra
+        if len(ccu.params.certificate) > 0 and len(trs.params.inboxUpdate) > 0:
+            certificate = decode(certificateSchema, ccu.params.certificate)
+            timestamp = timestamp of the block including ccu
+            if len(trs.params.inboxUpdate) > 0 and timestamp - certificate.timestamp > LIVENESS_LIMIT / 2:
+                raise Exception(f"First certificate cannot be older than {LIVENESS_LIMIT / 2} seconds.")
+``` 
 
-Those options for partial cross-chain update transactions are not expected to be used regularly. It is however a good option in corner case events where the sidechain outbox suddenly becomes very large.
+#### verifyCertificate
 
-#### Malicious Cross-chain Update Transactions
+```python
+def verifyCertificate(ccu: CCU) -> None:
+    certificateBytes = ccu.params.certificate      
+    certificate = decode(certificateSchema, certificateBytes)
+    partnerchainID = ccu.params.sendingChainID
+    partnerchainAccount = chainAccount(partnerchainID)
+    # Certificate height must be greater than last certificate height.
+    if certificate.height <= partnerchainAccount.lastCertificate.height:
+        raise Exception("Certificate height is not greater than last certificate height.")
+    # Certificate timestamp must be smaller than timestamp of the block including the CCU.
+    timestamp = timestamp of the block including ccu
+    if certificate.timestamp >= timestamp:
+        raise Exception("Certificate timestamp is not smaller than timestamp of the block including the CCU.")
+```
 
-A cross-chain update transaction may have been created with validly signed information according to the [validity rules](#parameters-validity) and hence be included in the blockchain. However, the transaction could include invalid CCMs. Those will be detected when trying to process the `inboxUpdate`.
+#### verifyCertificateSignature
 
-If a CCM is invalid (as specified in the ["Execute Cross-chain Updates" section](#execute-cross-chain-updates)), then the sending chain is [terminated][lip-0045#termination].
-The CCU processing then continues and CCMs from other chains are applied.
-All CCMs from terminated chains have no effect, they are neither applied nor forwarded.
+```python
+def verifyCertificateSignature(ccu: CCU) -> None:
+    # Certificate signature must be a valid aggregate signature for the sdechain validators.
+    certificate = decode(certificateSchema, ccu.params.certificate)
+    partnerchainID = ccu.params.sendingChainID
+    partnerchainAccount = chainAccount(partnerchainID)
+    validatorKeys = [validator.blsKey for validator in validators(partnerchainID).activeValidators]
+    validatorWeights = [validator.bftWeight for validator in validators(partnerchainID).activeValidators]
+    verifyAggregateCertificateSignature(
+        validatorKeys, 
+        validatorWeights, 
+        validators(partnerchainID).certificateThreshold, 
+        partnerchainID, 
+        certificate
+    )
+```
```

Comment:
Note that the function `verifyAggregateCertificateSignature` returns a boolean so we still need to check the return value and raise an exception instead of an invalid certificate.

Reply:
We agreed to update that function to actually raise an error.

Reply:
I did not remember that. All our verify functions such as `verifyWeightedAggSig`, `verifyBLS`, `verifyEd25519` return a Boolean at the moment so for consistency I think I would prefer to keep the function as is.

Reply:
fixed